marklogic-client-api/src/test/java/com/marklogic/client/test/dataservices/CallbackGeneratorTest.java

```diff
@@ -0,0 +1,415 @@
+package com.marklogic.client.test.dataservices;
+
+import static org.hamcrest.core.StringStartsWith.startsWith;
+import static org.junit.Assert.assertNotNull;
+
+import java.io.File;
+import java.io.InputStream;
+import java.io.Reader;
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetDateTime;
+import java.time.OffsetTime;
+import java.util.Date;
+
+import javax.xml.stream.XMLEventReader;
+import javax.xml.stream.XMLStreamReader;
+import javax.xml.transform.Source;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.w3c.dom.Document;
+import org.xml.sax.InputSource;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.marklogic.client.DatabaseClient;
+import com.marklogic.client.MarkLogicInternalException;
+import com.marklogic.client.dataservices.CallBatcher.CallArgsGenerator;
+import com.marklogic.client.dataservices.CallManager;
+import com.marklogic.client.dataservices.CallManager.CallArgs;
+import com.marklogic.client.dataservices.impl.CallBatcherImpl;
+import com.marklogic.client.dataservices.impl.CallBatcherImpl.BuilderImpl;
+import com.marklogic.client.dataservices.impl.CallManagerImpl.CallerImpl;
+import com.marklogic.client.io.marker.AbstractWriteHandle;
+import com.marklogic.client.test.Common;
+
+public class CallbackGeneratorTest implements CallArgsGenerator<CallManager.CallEvent>{
+
+    private static DatabaseClient db = Common.connect();
+    private static CallerImpl callerImpl;
+    private static CallbackGeneratorTest callbackGeneratorTest;
+    
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    @BeforeClass
+    public static void setup() {
+        callerImpl = new CallerImpl<CallManager.CallEvent>(null) {
```

Comment:
It's best if the tests use only the public interfaces.  That way, the implementation can change freely without destabilizing the interface.  (That's one motivation for separating the interface and implementation.)

In other words, the tests should create a CallManager, caller, args, CallBatcherBuilder, and CallBatcher in the same way that a developer will be creating them.  That way, we are confirming that the interfaces that the developer uses will work correctly.

The tests should execute a call to an endpoint on the enode and do something with the response.

One possibility would be to create an endpoint that takes and returns a single double.  The test callback could return 1.1 when passed a null event on the first call and -- on subsequent calls -- save the returned value in a final set and (if the value is less than 10) modify the args by incrementing the value or (if greater than 10) returning null.

After stopping the job, the test could then confirm that the actual set is the same as the expected set.